docs/code-standards-policy.rst

```diff
@@ -118,36 +171,418 @@ using the debops.apt_install_ role MUST be defined as (or similar):
+the firewall role which access rules to configure. This is done via dependent
+variables.
+
+Each role providing a feature which MAY be consumed by other roles MUST provide
+a dedicated dependent variable for the configuration of this feature. As that
+variable is always defined in the context of the consuming role, its value MUST
+NOT modify a shared configuration state (e.g. configuration template). The
+provider role might be executed multiple times depending on the role dependency
+configuration of the consuming roles.
+
+.. caution::
+
+   A role providing dependent variables MUST be able to handle multiple
+   role executions with different values of the dependent variables without
+   mutual interference.
+
+Ideally the dependent variable SHOULD accept a list of YAML dictionaries
+where one property MUST be the state (``present`` or ``absent``) of the
+configuration. The provider role SHOULD then manage an individual configuration
+file per list item which allows it to selectively add or remove configuration
+states.
+
+**Example:**
+
+The debops.apt_preferences_ role is implements a feature to set APT package
+pinning configurations. Each role requiring a specific version of a package
+to be available can as the ``apt_preferences`` to do the corresponding
+configuration. For this, ``apt_preferences`` provides the following dependent
+variable:
 
 .. code-block:: yaml
 
-   - name: Install the application
-     hosts: 'application_hosts'
+   # List of :manpage:`apt_preferences(5)` pins to configure in
+   # :file:`/etc/apt/preferences.d/`.  This variable is meant to be used from a
+   # role dependency in :file:`role/meta/main.yml` or in a playbook.
+   apt_preferences__dependent_list: []
+
+The consumer role, in this case debops.nginx_ would then define an own variable
+which defines the necessary pinning information:
+
+.. code-block:: yaml
+
+   nginx__apt_preferences__dependent_list:
+     - package: 'nginx nginx-*'
+       backports: [ 'wheezy', 'precise' ]
+
+In the playbook a :ref:`soft dependency <debops_policy__ref_code_standards_soft_role_dependencies>`
+can be specified where the dependent variable is passed to the provider:
+
+.. code-block:: yaml
+
+   - name: Manage nginx webserver
+     hosts: 'debops_service_nginx'
      become: True
 
      roles:
 
        - role: debops.apt_preferences
          apt_preferences__dependent_list:
-           - '{{ application__apt_preferences__dependent_list }}'
+           - '{{ nginx__apt_preferences__dependent_list }}'
 
-       - role: application
+       - role: debops.nginx
 
-By including the configuration for the debops.apt_preferences_ role in your role's
-default variables you allow the user to change it through the Ansible inventory
+By including the configuration for the debops.apt_preferences_ role in the
+``nginx`` default variables the user to change it through the Ansible inventory
 without the need to modify any of the involved roles or the playbook.
```

Comment:
Hmn, not sure if this is a good example that you mention here. If a dependent variable is properly used, it shouldn't change the functionality of the provider role for other consumers.

In my opinion "global" properties mustn't be set via dependent variables, exactly because conflicts cannot be avoided. If role A sets `nginx` version to 1 and role B to 2, the version would always be changed depending on the context in which it is run. I figured out, that in situation like this, you would simply "lock" the version via local facts, so that it cannot be changed anymore after the initial definition, but I don't fancy this feature so much. That definitely will result in troubles sooner or later if the roles don't behave anymore in the way how you configure them...